rmw_opensplice_cpp/src/topic_cache.hpp

```diff
@@ -0,0 +1,214 @@
+// Copyright 2014-2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TOPIC_CACHE_HPP_
+#define TOPIC_CACHE_HPP_
+
+
+#include <utility>
+#include <set>
+#include <string>
+#include <map>
+#include <sstream>
+#include <iterator>
+#include <algorithm>
+#include <mutex>
+#include "rcutils/logging_macros.h"
+
+
+/**
+ * Topics to types.
+ */
+typedef std::map<std::string, std::set<std::string>> TopicsTypes;
+
+/**
+ * Topic cache data structure.
+ * Manages relationships between participants and topics.
+ */
+template<typename GUID_t>
+class TopicCache
+{
+private:
+  /**
+   * Relevant Topic information for building relationship cache.
+   */
+  struct TopicInfo
+  {
+    GUID_t participant_guid;
+    GUID_t topic_guid;
+    std::string name;
+    std::string type;
+  };
+
+  typedef std::map<GUID_t, std::multiset<GUID_t>> ParticipantToTopicGuidMap;
+  typedef std::map<GUID_t, TopicInfo> TopicGuidToInfo;
+
+  /**
+   * Map of topic guid to topic info.
+   * Topics here are represented as one to many, DDS XTypes 1.2
+   * specifies application code 'generally' uses a 1-1 relationship.
+   * However, generic services such as logger and monitor, can discover
+   * multiple types on the same topic.
+   *
+   */
+  TopicGuidToInfo topic_guid_to_info_;
+
+  /**
+   * Map of participant GUIDS to a set of topic-type.
+   */
+  ParticipantToTopicGuidMap participant_to_topic_guids_;
+
+  /**
+   * Helper function to initialize the set inside a participant map.
+   *
+   * @param map
+   * @param participant_guid
+   */
+  void InitializeParticipantMap(
```

Comment:
For consistency with the rest of the package, use `snake_case` for functions and methods.

Reply:
My mistake, internal coding standards escaped me. Happy to oblige!